netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -423,7 +839,479 @@ net_ebpf_extension_sock_addr_authorize_connection_classify(
+    }
+
+    return status;
+}
+
+/*
+ * Default action is BLOCK. If this callout is being invoked, it means at least one
+ * eBPF program is attached. Hence no connection should be allowed unless allowed by
+ * the eBPF program.
+ *
+ * If the eBPF program verdict is BLOCK, no context is created and AUTH classify will
+ * block the connection by default.
+ */
+void
+net_ebpf_extension_sock_addr_redirect_connection_classify(
+    _In_ const FWPS_INCOMING_VALUES* incoming_fixed_values,
+    _In_ const FWPS_INCOMING_METADATA_VALUES* incoming_metadata_values,
+    _Inout_opt_ void* layer_data,
+    _In_opt_ const void* classify_context,
+    _In_ const FWPS_FILTER* filter,
+    uint64_t flow_context,
+    _Inout_ FWPS_CLASSIFY_OUT* classify_output)
+{
+    uint32_t verdict;
+    NTSTATUS status = STATUS_SUCCESS;
+    net_ebpf_extension_sock_addr_wfp_filter_context_t* filter_context = NULL;
+    net_ebpf_extension_hook_client_t* attached_client = NULL;
+    bpf_sock_addr_t* sock_addr_ctx = NULL;
+    bpf_sock_addr_t* sock_addr_ctx_original = NULL;
+    uint32_t compartment_id = UNSPECIFIED_COMPARTMENT_ID;
+    FWPS_CONNECTION_REDIRECT_STATE redirect_state;
+    HANDLE redirect_handle;
+    uint64_t classify_handle = 0;
+    net_ebpf_extension_connection_context_t* connection_context_original = NULL;
+    net_ebpf_extension_connection_context_t* connection_context_redirected = NULL;
+    bool redirected = false;
+    FWP_ACTION_TYPE action = FWP_ACTION_BLOCK;
+    bool classify_handle_acquired = false;
+    bool v4_mapped = false;
+
+    UNREFERENCED_PARAMETER(layer_data);
+    UNREFERENCED_PARAMETER(flow_context);
+
+    if ((classify_output->rights & FWPS_RIGHT_ACTION_WRITE) == 0) {
+        // Do not modify anything and bail.
+        return;
+    }
+
+    sock_addr_ctx = (bpf_sock_addr_t*)ExAllocatePoolUninitialized(
+        NonPagedPoolNx, sizeof(bpf_sock_addr_t), NET_EBPF_EXTENSION_POOL_TAG);
+    if (sock_addr_ctx == NULL) {
+        status = STATUS_INSUFFICIENT_RESOURCES;
+        goto Exit;
+    }
+    sock_addr_ctx_original = (bpf_sock_addr_t*)ExAllocatePoolUninitialized(
+        NonPagedPoolNx, sizeof(bpf_sock_addr_t), NET_EBPF_EXTENSION_POOL_TAG);
+    if (sock_addr_ctx_original == NULL) {
+        status = STATUS_INSUFFICIENT_RESOURCES;
+        goto Exit;
+    }
+    memset(sock_addr_ctx, 0, sizeof(bpf_sock_addr_t));
+    memset(sock_addr_ctx_original, 0, sizeof(bpf_sock_addr_t));
+
+    _net_ebpf_extension_sock_addr_copy_wfp_connection_fields(incoming_fixed_values, sock_addr_ctx);
+    *sock_addr_ctx_original = *sock_addr_ctx;
+
+    // Check if this call is intended for us.
+    filter_context = (net_ebpf_extension_sock_addr_wfp_filter_context_t*)filter->context;
+    ASSERT(filter_context != NULL);
+    if (filter_context == NULL) {
+        status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+
+    // If the callout is invoked for v4, then it is safe to simply invoke the eBPF
```

Comment:
encapsulate in a static function.

Reply:
Not sure if it will be worth it. This block of code has only 2 main checks and is also setting 2 variables which are used later in the function -- `v4_mapped` and `action`.